src/server/data_source.rs

```diff
@@ -76,12 +79,16 @@ impl DataSource {
             };
             let DataSource { read, .. } = data_source;
 
-            // TODO: Prepare callback context.
-            let mut context = DataSourceReadContext;
+            let mut read = AssertUnwindSafe(read);
+            let mut context = DataSourceReadContext { value };
 
-            match read(&mut context) {
-                Ok(()) => ua::StatusCode::GOOD,
-                Err(status_code) => status_code,
+            match catch_unwind(move || read(&mut context)) {
```

Comment:
I remember this strategy from Tokio to avoid poisoning a lock if a callback panics.

I didn't mention it before, because we don't lock here. What would happen otherwise if the panic escapes? Maybe we should add a comment that explains the implications and the decision for catching it.